File: solutions/2026-09-01/cpp/2026-09-01-minimum-moves-to-clean-the-classroom.cpp
/*
Problem Summary:
This problem asks for the minimum number of moves a student needs to collect all litter ('L') in a grid classroom, starting from 'S'. The student has a maximum energy capacity and loses 1 energy per move. Energy can be restored to full at 'R' (reset) cells. Obstacles 'X' cannot be passed.

Link to the problem:
https://leetcode.com/problems/minimum-moves-to-clean-the-classroom/

Approach Explanation:
This problem can be modeled as a shortest path problem on a graph where nodes represent states and edges represent moves. Since we need the *minimum* number of moves, Breadth-First Search (BFS) is the appropriate algorithm.

A state in our BFS needs to capture:
1.  `row`: The current row position of the student.
2.  `col`: The current column position of the student.
3.  `current_energy`: The remaining energy the student has.
4.  `litter_mask`: A bitmask representing which litter items have been collected. Each 'L' cell is assigned a unique index (0 to `num_litter - 1`), and the corresponding bit in the mask is set to 1 when that litter is collected.

The BFS explores states layer by layer, ensuring that the first time we reach a state where all litter is collected, we have found the minimum number of moves.

The `dist[row][col][energy][litter_mask]` 4D array is used to store the minimum moves required to reach a specific state. It also serves as a `visited` set to prevent redundant computations and cycles. It's initialized to -1 (unvisited).

Energy mechanics:
- Moving to an adjacent cell costs 1 energy.
- If the student is on an 'R' cell, their energy is fully restored to `initial_energy` *before* they make the next move.
- If the student's `current_energy` is 0 and they are *not* on an 'R' cell, they cannot make any further moves from that position.
- If the student moves to an 'R' cell, their energy is immediately reset to `initial_energy` *upon arrival* at that cell.
- If the student moves to an 'L' cell, that litter is considered collected, and the `litter_mask` is updated.

Time Complexity:
The total number of states is `m * n * (energy + 1) * (1 << num_litter)`.
- `m, n` are grid dimensions (up to 20).
- `energy` is max energy (up to 50).
- `num_litter` is the count of 'L' cells (up to 10).
For each state, we explore at most 4 neighbors.
So, the time complexity is `O(m * n * energy * 2^num_litter)`.
With the given constraints: `20 * 20 * 50 * 2^10 = 400 * 50 * 1024 = 20,480,000` operations, which is efficient enough for typical time limits.

Space Complexity:
The primary space consumer is the `dist` array.
Its size is `m * n * (energy + 1) * (1 << num_litter)` integers.
`20 * 20 * 51 * 1024 * sizeof(int)` (assuming `sizeof(int) = 4` bytes).
`400 * 51 * 1024 * 4 = 83,558,400` bytes, which is approximately `83.5 MB`. This should be acceptable within typical memory limits (e.g., 256MB or 512MB). The queue also stores states, but its maximum size is bounded by the total number of states.
*/

#include <vector>   // For std::vector
#include <string>   // For std::string
#include <queue>    // For std::queue
#include <tuple>    // For std::tuple
#include <cstring>  // For memset

// Define maximum grid dimensions, energy, and litter count based on constraints.
const int MAX_M = 20;
const int MAX_N = 20;
const int MAX_ENERGY = 50;
const int MAX_LITTER_COUNT = 10; // Maximum number of 'L' cells allowed

// The dist array stores the minimum moves to reach a specific state (r, c, current_energy, litter_mask).
// Initialized to -1 to signify unvisited states.
// Declared globally or as static member to avoid stack overflow for large fixed-size arrays.
int dist[MAX_M][MAX_N][MAX_ENERGY + 1][1 << MAX_LITTER_COUNT];

// litter_index_grid stores the assigned index for each 'L' cell.
// If a cell is not 'L', it stores -1. This helps quickly get the bit index for the mask.
int litter_index_grid[MAX_M][MAX_N];

// Arrays for movement directions: up, down, left, right.
int dr[] = {-1, 1, 0, 0};
int dc[] = {0, 0, -1, 1};

class Solution {
public:
    int minimumMoves(std::vector<std::string>& classroom, int energy) {
        // Get the dimensions of the classroom grid.
        int m = classroom.size();
        int n = classroom[0].size();
        
        // Initialize the dist array with -1 to mark all states as unvisited.
        // memset is efficient for initializing large C-style arrays with a specific byte value (like -1 or 0).
        memset(dist, -1, sizeof(dist));
        
        // Variables to store the starting position and process litter locations.
        int start_r, start_c;
        std::vector<std::pair<int, int>> litter_locations; // Stores (r, c) for each 'L'

        // Initialize litter_index_grid with -1 for all cells.
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                litter_index_grid[i][j] = -1;
            }
        }

        // Iterate through the grid to find 'S' and all 'L' cells.
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (classroom[r][c] == 'S') {
                    start_r = r; // Store starting row
                    start_c = c; // Store starting column
                } else if (classroom[r][c] == 'L') {
                    // Assign a unique index to each 'L' cell based on its order of discovery.
                    litter_index_grid[r][c] = litter_locations.size();
                    litter_locations.push_back({r, c}); // Store coordinates
                }
            }
        }
        
        // Get the total number of litter items.
        int num_litter = litter_locations.size();
        // Calculate the target bitmask: all bits set from 0 to num_litter-1.
        // E.g., if num_litter = 3, (1 << 3) - 1 = 8 - 1 = 7 (binary 111).
        int all_litter_collected_mask = (1 << num_litter) - 1;
        
        // Initialize BFS queue. Each element is a tuple: (row, col, current_energy, litter_mask).
        std::queue<std::tuple<int, int, int, int>> q;
        
        // Push the initial state into the queue.
        // Starting position, initial max energy, 0 litter collected.
        q.push({start_r, start_c, energy, 0});
        // Mark the initial state as visited and set its distance to 0 moves.
        dist[start_r][start_c][energy][0] = 0;
        
        // Main BFS loop.
        while (!q.empty()) {
            // Dequeue the current state.
            auto [r, c, current_energy, current_mask] = q.front();
            q.pop();
            
            // Get the minimum moves taken to reach this current state.
            int moves = dist[r][c][current_energy][current_mask];
            
            // If all litter items have been collected, we found the shortest path.
            if (current_mask == all_litter_collected_mask) {
                return moves;
            }
            
            // Determine the effective energy available for making a move *from* the current cell.
            // If the current cell is 'R', the student's energy is reset to full capacity *before* moving.
            // Otherwise, they use their current remaining energy.
            int effective_energy_for_move_from_current_cell;
            if (classroom[r][c] == 'R') {
                effective_energy_for_move_from_current_cell = energy; // Full initial energy
            } else {
                effective_energy_for_move_from_current_cell = current_energy;
            }

            // If the effective energy is 0, the student cannot make any further moves from this cell.
            // This condition is important as it implies energy must be positive to move.
            if (effective_energy_for_move_from_current_cell == 0) {
                continue;
            }

            // Explore all four adjacent cells (up, down, left, right).
            for (int i = 0; i < 4; ++i) {
                int next_r = r + dr[i];
                int next_c = c + dc[i];
                
                // Check if the next cell is within grid boundaries.
                if (next_r < 0 || next_r >= m || next_c < 0 || next_c >= n) {
                    continue; // Out of bounds
                }
                
                // Check if the next cell is an obstacle ('X').
                if (classroom[next_r][next_c] == 'X') {
                    continue; // Cannot move through obstacles
                }
                
                // Calculate the energy after moving to the next cell.
                // Each move costs 1 unit of energy.
                int next_energy = effective_energy_for_move_from_current_cell - 1;
                // Initialize the litter mask for the next state; it's initially the same as current.
                int next_mask = current_mask;
                
                // Apply specific effects based on the type of the destination cell (next_r, next_c).
                if (classroom[next_r][next_c] == 'L') {
                    // If moving to an 'L' cell, collect the litter by setting its corresponding bit in the mask.
                    next_mask |= (1 << litter_index_grid[next_r][next_c]);
                }
                
                if (classroom[next_r][next_c] == 'R') {
                    // If moving to an 'R' cell, the energy is reset to full capacity *upon arrival*.
                    next_energy = energy;
                }
                
                // Check if this new state (next_r, next_c, next_energy, next_mask) has been visited before,
                // or if we found a shorter path to it.
                if (dist[next_r][next_c][next_energy][next_mask] == -1) {
                    // If unvisited, mark it with the new minimum moves and enqueue it.
                    dist[next_r][next_c][next_energy][next_mask] = moves + 1;
                    q.push({next_r, next_c, next_energy, next_mask});
                }
            }
        }
        
        // If the BFS queue becomes empty and we haven't returned (meaning all litter was not collected),
        // it's impossible to collect all litter.
        return -1;
    }
};

File: solutions/2026-09-01/python/2026-09-01-minimum-moves-to-clean-the-classroom.py
# Problem: Minimum Moves to Clean the Classroom
# Summary: Find the minimum moves to collect all litter ('L') in a grid, starting from 'S', with limited energy that can be replenished at 'R' cells.
# Link: https://leetcode.com/problems/minimum-moves-to-clean-the-classroom/
# Approach: This problem can be modeled as a shortest path problem on a state graph. Each state is defined by (row, col, current_energy, collected_litter_mask). Since we need to find the minimum number of moves, Breadth-First Search (BFS) is appropriate.
#
# The state for BFS will be (row, col, current_energy, collected_litter_mask).
# - `row`, `col`: Current position in the grid.
# - `current_energy`: Remaining energy. This can range from 0 to `initial_energy`.
# - `collected_litter_mask`: A bitmask representing which litter items have been collected. Since there are at most 10 'L' cells, a 10-bit integer is sufficient (2^10 = 1024 states).
#
# We need to pre-process the grid to find the coordinates of all 'L' cells and assign them an index (0 to N-1) to build the bitmask.
#
# BFS Steps:
# 1. Initialize a queue with the starting state: (start_row, start_col, initial_energy, 0, 0_moves).
# 2. Use a `visited` set to store (row, col, current_energy, collected_litter_mask) to avoid redundant computations and cycles.
# 3. When exploring neighbors:
#    a. A move costs 1 energy. If energy becomes 0 and the cell is not 'R', it's an invalid move.
#    b. If the cell is 'R', reset energy to `initial_energy`.
#    c. If the cell contains 'L', update the `collected_litter_mask`.
#    d. If the new state (neighbor_row, neighbor_col, new_energy, new_mask) has not been visited, add it to the queue and mark as visited.
# 4. The BFS terminates when a state is reached where `collected_litter_mask` indicates all litter items have been collected. The number of moves at that point is the answer.
# 5. If the queue becomes empty and not all litter items are collected, it's impossible, return -1.
#
# Time Complexity:
# Let M be the number of rows, N be the number of columns.
# Let E be the initial maximum energy.
# Let L be the number of litter cells (at most 10).
#
# The number of possible states is `M * N * (E + 1) * 2^L`.
# - `M * N`: For grid positions.
# - `E + 1`: For energy levels (0 to E).
# - `2^L`: For the litter collection mask.
#
# Each state can have up to 4 neighbors. Processing each state involves constant time operations.
# So, the time complexity is approximately O(M * N * E * 2^L).
# Given M, N <= 20, E <= 50, L <= 10:
# 20 * 20 * 50 * 2^10 = 400 * 50 * 1024 = 20000 * 1024 = 20,480,000.
# This should be acceptable within typical time limits (usually ~10^8 operations).
#
# Space Complexity:
# The `visited` set and the BFS queue can store up to O(M * N * E * 2^L) states.
# Each state stores (row, col, energy, mask, moves).
# So, the space complexity is O(M * N * E * 2^L).
# Using the same numbers, 20,480,000 states, each state a tuple, will consume significant memory but likely within limits (e.g., 20M tuples of 4 integers).

import collections

class Solution:
    def minimumMoves(self, classroom: list[str], energy: int) -> int:
        m, n = len(classroom), len(classroom[0])

        # Find starting position 'S' and all 'L' positions
        start_pos = None
        litter_coords = []
        for r in range(m):
            for c in range(n):
                if classroom[r][c] == 'S':
                    start_pos = (r, c)
                elif classroom[r][c] == 'L':
                    litter_coords.append((r, c))
        
        # Map litter coordinates to an index for the bitmask
        litter_map = {coord: i for i, coord in enumerate(litter_coords)}
        total_litter = len(litter_coords)
        
        # Target mask when all litter is collected
        target_mask = (1 << total_litter) - 1

        # BFS Queue: stores (row, col, current_energy, collected_litter_mask, moves)
        # Using a deque for efficient popleft
        queue = collections.deque()

        # Visited set: stores (row, col, current_energy, collected_litter_mask)
        # We don't store moves here because BFS inherently finds the shortest path,
        # so if we visit a state with more moves, it's not optimal.
        visited = set()

        # Add initial state to queue
        # (start_row, start_col, initial_energy, 0_collected_mask, 0_moves)
        start_r, start_c = start_pos
        queue.append((start_r, start_c, energy, 0, 0))
        visited.add((start_r, start_c, energy, 0))

        # Directions for movement: up, down, left, right
        dr = [-1, 1, 0, 0]
        dc = [0, 0, -1, 1]

        while queue:
            r, c, current_energy, current_mask, moves = queue.popleft()

            # If all litter collected, return moves
            if current_mask == target_mask:
                return moves

            # Explore all 4 adjacent cells
            for i in range(4):
                nr, nc = r + dr[i], c + dc[i]

                # Check bounds
                if not (0 <= nr < m and 0 <= nc < n):
                    continue

                # Obstacle 'X' cannot be passed
                if classroom[nr][nc] == 'X':
                    continue

                # Calculate new energy after moving
                new_energy = current_energy - 1

                # If energy drops to 0 or below, check if we are on 'R'
                # If new_energy is negative, it implies we moved without sufficient energy.
                # However, the problem states "If the energy reaches 0, the student can only continue
                # if they are on a reset area 'R'". This implies if current_energy is 0 and we
                # move to a non-'R' cell, it's invalid.
                # If current_energy is > 0 and we move to a cell, new_energy becomes current_energy - 1.
                # If current_energy is 1, new_energy becomes 0. If this cell is not 'R', we can't move from it.
                # This logic is applied at the point of making the move.
                # Let's simplify: moving costs 1 energy. If current_energy is 0, we can't move unless
                # we are standing on 'R' (which resets energy *before* moving).
                # The interpretation should be: if current_energy - 1 < 0 and the current cell is not 'R',
                # then we cannot move.
                # A better way to handle energy:
                # 1. Moving always costs 1 energy.
                # 2. If new_energy < 0, this move is impossible.
                # 3. If the destination cell classroom[nr][nc] is 'R', then new_energy becomes `energy` (max capacity).
                #
                # Re-evaluating the energy rule: "If the energy reaches 0, the student can only continue if they are on a reset area 'R', which resets the energy to its maximum capacity energy."
                # This implies:
                # - If current_energy > 0: Student can move. New energy will be current_energy - 1.
                # - If current_energy == 0: Student *cannot* move to any cell unless the *current cell* (r, c) is 'R'.
                #   If (r, c) is 'R', energy gets reset to `energy` *before* the move, and then a move costs 1.
                #   So, from (r, c) being 'R' with 0 energy, student can move with `energy - 1`.
                
                # Case 1: Student has energy > 0.
                if current_energy > 0:
                    next_energy_after_move = current_energy - 1
                    next_mask = current_mask
                    
                    # If destination is 'L', collect it
                    if (nr, nc) in litter_map:
                        litter_idx = litter_map[(nr, nc)]
                        next_mask |= (1 << litter_idx)
                    
                    # If destination is 'R', reset energy
                    if classroom[nr][nc] == 'R':
                        next_energy_after_move = energy # Reset to full capacity

                    # Check if this new state has been visited
                    if (nr, nc, next_energy_after_move, next_mask) not in visited:
                        visited.add((nr, nc, next_energy_after_move, next_mask))
                        queue.append((nr, nc, next_energy_after_move, next_mask, moves + 1))
                
                # Case 2: Student has 0 energy.
                # They can only move if current position (r, c) is an 'R' cell.
                elif current_energy == 0: # current_energy is 0
                    if classroom[r][c] == 'R':
                        # Energy is reset to max BEFORE moving
                        energy_after_reset_and_move = energy - 1
                        
                        # This implies we can only move if `energy - 1 >= 0`, which is true since energy >= 1.
                        # However, if `energy` itself is 0, this rule breaks. Constraints state `energy >= 1`.
                        
                        next_mask = current_mask
                        # If destination is 'L', collect it
                        if (nr, nc) in litter_map:
                            litter_idx = litter_map[(nr, nc)]
                            next_mask |= (1 << litter_idx)
                        
                        # If destination is 'R', reset energy. This is essentially double reset,
                        # first at (r,c), then at (nr,nc). The rule is "restores the student's energy to full capacity"
                        # when *on* 'R'. So if we move to 'R', energy becomes full.
                        # The `energy_after_reset_and_move` is energy - 1. If (nr,nc) is 'R', it becomes `energy`.
                        if classroom[nr][nc] == 'R':
                            next_energy_after_move_final = energy
                        else:
                            next_energy_after_move_final = energy_after_reset_and_move

                        if (nr, nc, next_energy_after_move_final, next_mask) not in visited:
                            visited.add((nr, nc, next_energy_after_move_final, next_mask))
                            queue.append((nr, nc, next_energy_after_move_final, next_mask, moves + 1))
                    # If current_energy is 0 and current cell is not 'R', student cannot move.
                    # This branch does nothing, which is correct.

        # If queue becomes empty and target_mask was not reached
        return -1

# Test cases (example usage):
# sol = Solution()
# print(sol.minimumMoves(classroom = ["S.", "XL"], energy = 2)) # Expected: 2
# print(sol.minimumMoves(classroom = ["LS", "RL"], energy = 4)) # Expected: 3
# print(sol.minimumMoves(classroom = ["L.S", "RXL"], energy = 3)) # Expected: -1